### PR TITLE
Address Ranges support (CIDR notation)

### PR DESCRIPTION
**Pull request by willtho89**
This PR allows the use of Address Ranges using the CIDR notation. 

To make it backward compatible, i introduced a new env variable WG_DEFAULT_ADDRESS_RANGE (defaults to the previous default of 24). 

This allows the usage of smaller subnets (or possibly larger; but i didn't test that due to restrictions on my network). Client IPs will be calculated with correct IP addresses instead of making assumptions of the address space.

Fixes: #140 
Closes: #393

### CODE REVIEW

File: src/lib/WireGuard.js [hunk 1]
```diff
@@ -199,7 +205,7 @@ AllowedIPs = ${client.address}/32`;
     return `
 [Interface]
 PrivateKey = ${client.privateKey}
-Address = ${client.address}/24
```
Comment (willtho89) [inline src/lib/WireGuard.js:202]: Not 100% sure if this was a wanted behavior. Previously the exported client config had the 24 subnet, even tough the wg.conf sets it to /32 (https://github.com/wg-easy/wg-easy/pull/797/files#diff-65b7aac2a6e96abf37ec0c4c0c24a4e94ac70fc6a0702d74f45f79b900716f79L114)

**Comment by pheiduck** [comment]
Nice, so User can change there IP range, if they want otherwise its on /24.
I like it!

File: src/lib/WireGuard.js (hunk 1 above)
Comment (pheiduck) [reply src/lib/WireGuard.js:202]: Address is not the same as AllowedIPs

**Comment by pheiduck** [comment]
Set to draft, because untested.

File: src/lib/WireGuard.js (hunk 1 above)
Comment (pheiduck) [reply src/lib/WireGuard.js:202]: This is the cidr for the client IP. It may therefore need to be replaced with your new variable.

Comment (willtho89) [reply src/lib/WireGuard.js:202]: I get that Address != AllowedIPs, but if the peer on the "server" has an allowed IP of /32 for the peer (as in line 114), it does not expect the "client" interface to have a subnet of /24. Or am i missing something?

Comment (pheiduck) [reply src/lib/WireGuard.js:202]: The keyword allowed-ips is a list of addresses that will get routed to the peer. Make sure to specify at least one address range that contains the WireGuard connection's internal IP address(es). from: https://wiki.archlinux.org/title/WireGuard#Peer_setup

Comment (willtho89) [reply src/lib/WireGuard.js:202]: Okay i did some more research into this. 
Even tough it shouldn't matter to much for the use case wg-easy provides, i will change this line back to 24. If some users have site-site networks they need the /24, for everyone else it should not matter.

**Comment by peterlewis** [comment]
I like this, thank you, but think it would work better and be cleaner if it were possible to use CIDR notation in the `WG_DEFAULT_ADDRESS` environment variable (e.g. `10.0.8.0/24`), as opposed to introducting a new variable to compliment the existing. Does that resonate with you, at all?

**Comment by willtho89** [comment]
> I like this, thank you, but think it would work better and be cleaner if it were possible to use CIDR notation in the `WG_DEFAULT_ADDRESS` environment variable (e.g. `10.0.8.0/24`), as opposed to introducting a new variable to compliment the existing. Does that resonate with you, at all?

Would be okay for me. 

I'll see if i can do that without messing up the current implementation ("x" as last char)

**Comment by willtho89** [comment]
@peterlewis i refactored my code to support both the legacy syntax and CIDR notation with WG_DEFAULT_ADDRESS

**Comment by peterlewis** [comment]
> @peterlewis i refactored my code to support both the legacy syntax and CIDR notation with WG_DEFAULT_ADDRESS

Thanks! Will endeavour to take a proper look tomorrow but one thing to point out is that CODR ranges typically end with a `0`, and would need to, in the context of how they are being used here.

**Comment by willtho89** [comment]
> > @peterlewis i refactored my code to support both the legacy syntax and CIDR notation with WG_DEFAULT_ADDRESS
> 
> Thanks! Will endeavour to take a proper look tomorrow but one thing to point out is that CIDR ranges typically end with a `0`, and would need to, in the context of how they are being used here.

You are absolutely correct. I will fix this once a review has been done 😊

File: README.md [hunk 2]
```diff
@@ -90,7 +90,7 @@ These options can be configured by setting environment variables using `-e KEY="
 | `WG_PORT` | `51820` | `12345` | The public UDP port of your VPN server. WireGuard will always listen on 51820 inside the Docker container. |
 | `WG_MTU` | `null` | `1420` | The MTU the clients will use. Server uses default WG MTU. |
 | `WG_PERSISTENT_KEEPALIVE` | `0` | `25` | Value in seconds to keep the "connection" open. If this value is 0, then connections won't be kept alive. |
-| `WG_DEFAULT_ADDRESS` | `10.8.0.x` | `10.6.0.x` | Clients IP address range. |
+| `WG_DEFAULT_ADDRESS` | `10.8.0.1/24` | `10.6.0.x` | Clients IP address range. (For Legacy reasons x in last place is supported (e.g. 10.8.0.x)) |
```
Comment (peterlewis) [inline README.md:93]: I note you've set the default to CIDR format (makes sense). The README will need updating to reflect this.

File: docker-compose.yml [hunk 3]
```diff
@@ -15,7 +15,7 @@ services:
       # Optional:
       # - PASSWORD=foobar123
       # - WG_PORT=51820
-      # - WG_DEFAULT_ADDRESS=10.8.0.x
+      # - WG_DEFAULT_ADDRESS=10.8.0.1/24
```
Comment (peterlewis) [inline docker-compose.yml:18]: `10.0.8.0/24`

File: src/config.js [hunk 4]
```diff
@@ -1,7 +1,29 @@
 'use strict';
 
+const ip = require('ip');
+
 const { release } = require('./package.json');
 
+function parseDefaultAddress(defaultAddress) {
+  // Set the default full address with subnet if it's not provided
+  const defaultFullAddress = defaultAddress || '10.8.0.1/24';
```
Comment (peterlewis) [inline src/config.js:9]: `10.0.8.0/24`

File: src/config.js [hunk 5]
```diff
@@ -1,7 +1,29 @@
...
+
+  // Check if the address ends with '.x', if so, replace with '.1/24'
+  const addressWithSubnet = defaultFullAddress.endsWith('.x')
+    ? defaultFullAddress.replace('.x', '.1/24')
```
Comment (peterlewis) [inline src/config.js:13]: `.0/24`

File: src/config.js [hunk 6]
```diff
@@ -1,7 +1,29 @@
...
+
+  const [ipAddress, subnetRange] = addressWithSubnet.split('/');
+
+  return {
+    ipAddress,
+    subnetRange: subnetRange || '24', // Default subnet range to 24 if not provided
+  };
+}
+
+// Use the function to parse the environment variable or default to '10.8.0.1/24'
```
Comment (peterlewis) [inline src/config.js:24]: `10.0.8.0/24`

File: wg-easy.service [hunk 7]
```diff
@@ -5,7 +5,7 @@ After=network-online.target nss-lookup.target
 [Service]
 Environment="WG_HOST=raspberrypi.local" # Change this to your host's public address or static public ip.
 Environment="PASSWORD=REPLACEME" # When set, requires a password when logging in to the Web UI, to disable add a hashtag
-#Environment="WG_DEFAULT_ADDRESS=10.0.8.x" #Clients IP address range.
+#Environment="WG_DEFAULT_ADDRESS=10.0.8.1/24" #Clients IP address range.
```
Comment (peterlewis) [inline wg-easy.service:8]: `10.0.8.0/24`

**Comment by peterlewis** [comment]
Hey @willtho89 it's looking promising, but realised my review is somewhat verbose. If you are please happy to make all required changes first and then I can re-review in greater detail that'd be great.

File: README.md (hunk 2 above)
Comment (willtho89) [reply README.md]: Not sure what you mean by this? Another paragraph describing the change? As we are fully backwards compatible, i would just mention it in the Changelog

**Comment by willtho89** [comment]
@peterlewis i made the requested changes

**Comment by pheiduck** [comment]
I would like to point out that there is a vulnerability in the 'ip' npm package. We should wait until this has been fixed. 

See: https://github.com/advisories/GHSA-78xj-cgh5-2h22

**Comment by peterlewis** [comment]
Thanks @pheiduck, agreed. Also keen to revisit https://github.com/wg-easy/wg-easy/pull/393 to see if there are any improvements/refinements in @BrainStone's previous PR that can be merged into this. Didn't spot the trigger-happy GitHub Actions bot closure! :-/

**Comment by pheiduck** [comment]
@willtho89 okay understand. Then we should provide it as base package. Due the docker image is not updated until now. See: https://hub.docker.com/layers/library/node/18-alpine/images/sha256-affdf979bd8ec516bf189d451b8ac68dd50adc49adc4c4014963556c11efeda4

**Comment by willtho89** [comment]
> @willtho89 okay understand. Then we should provide it as base package. Due the docker image is not updated until now. See: https://hub.docker.com/layers/library/node/18-alpine/images/sha256-affdf979bd8ec516bf189d451b8ac68dd50adc49adc4c4014963556c11efeda4
> 
> https://github.com/nodejs/node/blob/main/deps/npm/node_modules/ip/package.json#L3
> 
> Or wait until it is.

Its updated as of 2 hours ago https://www.npmjs.com/package/ip?activeTab=versions

I updated the package.json to use 2.0.1. Also I pulled the latest changes from this wg-easy:master. You might want to update the target as well (`wg-easy:feat-cidr-notation`).

**Comment by pheiduck** [comment]
@willtho89 Is everything ready here? I would include it in the [feat-cidr-notation](https://github.com/wg-easy/wg-easy/tree/feat-cidr-notation) branch for testing.

**Comment by willtho89** [comment]
Yep. The feature is already in use on our local company network and working as expected

**Review by peterlewis — APPROVED** [review]

**Comment by peterlewis** [comment]
Thanks, @willtho89. I'll get that build on `development`, now...

**Comment by peterlewis** [comment]
My `WG_DEFAULT_ADDRESS` is set to `172.16.2.x` and container errors on boot:

```
$ wg-quick up wg0
Error: Command failed: wg-quick up wg0
[#] 
[#] ip link add wg0 type wireguard
[#] wg setconf wg0 /dev/fd/63
[#] ip -4 address add 172.16.2.1/undefined dev wg0
Error: inet prefix is expected rather than "172.16.2.1/undefined".
[#] ip link delete dev wg0
    at ChildProcess.exithandler (node:child_process:422:12)
    at ChildProcess.emit (node:events:517:28)
    at maybeClose (node:internal/child_process:1098:16)
    at Socket.<anonymous> (node:internal/child_process:450:11)
    at Socket.emit (node:events:517:28)
    at Pipe.<anonymous> (node:net:350:12) {
  code: 1,
  killed: false,
  signal: null,
  cmd: 'wg-quick up wg0'
}
```

**Comment by pheiduck** [comment]
I guess it has to be `172.16.2.0/24`

**Comment by pheiduck** [comment]
It sadly break functionality for users they don't use this feature:
```
│   2024-02-24T22:04:51.911Z WireGuard Config saved.                                                                                                                             │
│   $ wg-quick down wg0                                                                                                                                                          │
│   $ wg-quick up wg0                                                                                                                                                            │
│   Error: Command failed: wg-quick up wg0                                                                                                                                       │
│   [#]                                                                                                                                                                          │
│   [#] ip link add wg0 type wireguard                                                                                                                                           │
│   [#] wg setconf wg0 /dev/fd/63                                                                                                                                                │
│   [#] ip -4 address add 10.8.0.1/undefined dev wg0                                                                                                                             │
│   Error: inet prefix is expected rather than "10.8.0.1/undefined".                                                                                                             │
│   [#] ip link delete dev wg0                                                                                                                                                   │
│   2024-02-24T22:04:52.208577463Z                                                                                                                                               │
│       at ChildProcess.exithandler (node:child_process:422:12)                                                                                                                  │
│       at ChildProcess.emit (node:events:517:28)                                                                                                                                │
│       at maybeClose (node:internal/child_process:1098:16)                                                                                                                      │
│       at ChildProcess._handle.onexit (node:internal/child_process:303:5) {                                                                                                     │
│     code: 1,                                                                                                                                                                   │
│     killed: false,                                                                                                                                                             │
│     signal: null,                                                                                                                                                              │
│     cmd: 'wg-quick up wg0'                                                                                                                                                     │
│   }
```

**Comment by peterlewis** [comment]
> I guess it has to be `172.16.2.0/24`

Yeah sorry that was a typo in my comment.

**Comment by willtho89** [comment]
@pheiduck @peterlewis unfortunately i can not reproduce this. Running ` docker compose -f docker-compose.yml -f docker-compose.dev.yml up` with either no `WG_DEFAULT_ADDRESS`, `WG_DEFAULT_ADDRESS=172.16.2.0/24` or `WG_DEFAULT_ADDRESS=172.16.2.x` the container comes up and i can connect to it.

**Comment by pheiduck** [comment]
@willtho89 thank you for this hint I will test it later.

**Comment by pheiduck** [comment]
@willtho89 okay I got the issue due the wg0.conf gets overwritten, the CIDR will get: 

```
Address = 10.8.0.1/undefined
```

**Pull request merged**